### PR TITLE
Add built-in "Open With" plugin

### LINKED ISSUE #705: Add "Open With" option to the editor
Would be nice if Kap had an "Open With" option in the Editor that would let me open the exported recording in any installed app. That way we could support other apps, even if they don't have a plugin integration. For example, I could export a recording, and then have it opened in Final Cut Pro.

It would look similar to the "Open With" menu in the Finder context menu.

It would be another item here:

Screenshot 2019-08-13 at 22 23 00

And be a submenu that lists apps able to handle the file type. We would have to make a binary module for this, but it's not hard:

```swift
extension NSWorkspace {
	func urlsForAppsThatCanOpen(typeIdentifier: String, role: LSRolesMask = [.viewer, .editor]) -> [URL] {
		let returnValue = LSCopyAllRoleHandlersForContentType(typeIdentifier as CFString, role)?.takeUnretainedValue() as? [String] ?? []
		return returnValue.compactMap { urlForApplication(withBundleIdentifier: $0) }
	}
}
```

And we would need that binary module to also handle opening a specific app: https://developer.apple.com/documentation/appkit/nsworkspace/1528548-open

### PR DESCRIPTION
**Pull request by karaggeorge**
Closes #705 

Also finally fixes https://github.com/wulkano/kap/pull/686#discussion_r301158258

Please test this a lot with exporting with every available option, (un)installing plugins while editor is open and exporting. Basically try to break it in any way. It contains a decent amount of refactor, so I want to make sure it didn't break anything

Built this module for this. Might need some review as well: https://github.com/karaggeorge/mac-open-with

### CODE REVIEW

**Comment by sindresorhus** [comment]
macOS 10.15 returns the apps in the correct order, so can you conditionalize https://github.com/karaggeorge/mac-open-with/blob/6bd752b2e10c209043651ab37499430f40745c47/index.js#L13-L17 to only do that on macOS 10.14 and below?

**Comment by sindresorhus** [comment]
You need to filter out Kap and Kap Beta:

Screenshot 2019-09-12 at 11 57 29

File: main/plugins/open-with-plugin.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+const apps = new Map(
+  ['mp4', 'gif', 'apng', 'webm']
+    .map(extension => [
+      extension,
+      getAppsThatOpenExtension.sync(extension).map(app => ({
+        ...app,
+        name: decodeURI(path.parse(app.url).name)
+      }))
+    ])
+    .filter(([_, apps]) => apps.length > 0)
```
Comment (sindresorhus) [inline main/plugins/open-with-plugin.js]: Can you sort `Gifski` to the top? People are likely to want to use it with Kap if they have it installed.

**Comment by sindresorhus** [comment]
You should not show the default in the normal list:

Screenshot 2019-09-12 at 12 06 12

Notice how Chromium is in the list twice.

**Comment by sindresorhus** [comment]
There seems to be no proper sorting of plugins now:

Screenshot 2019-09-12 at 12 10 12
Screenshot 2019-09-12 at 12 10 19

I think "Save to Disk" should be first, then "Copy to Clipboard, then "Open With" and then third-party plugins.

**Comment by sindresorhus** [comment]
> There seems to be no proper sorting of plugins now:

Ah, I forgot, we sort on usage, right?

**Comment by sindresorhus** [comment]
@karaggeorge You need to guard this plugin behind a check. The binary will only work on macOS 10.14.4 and later as the Swift libraries are now bundled in the OS and not in the binary. See: https://github.com/karaggeorge/mac-open-with/pull/1/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9R5

**Comment by karaggeorge** [comment]
Hmm, I'll look into that. Also, I tried to check it but it seems like when the role is `submenu` and not one of `checkbox` or `radio` you can't check it (at least with the Electron API, not sure natively)

**Comment by karaggeorge** [comment]
@sindresorhus should be fixed

**Comment by sindresorhus** [comment]
Looks good :)

**Comment by sindresorhus** [comment]
> Also, I tried to check it but it seems like when the role is submenu and not one of checkbox or radio you can't check it (at least with the Electron API, not sure natively)

You can do it natively, but not very important.

**Pull request merged**